### ISSUE TITLE
UI Freezing on visual notifications

### ISSUE THREAD

**Issue by zd-zero** [zd-zero/waifu-motivator-plugin#206]
I noticed this when I test the different events of the plugin and sometimes the platform is freezing, the culprit is the response time of retrieving the assets. We should bundle at least 3 visual notifications that the plugin can fallback into whenever the asset retrieval takes a while, this can also help if the user is in offline mode.

The idea is, we need to request download assets if it returns fast enough, use it, else use the fallback.
Fallback order whenever it takes a while to download:
If the download takes a while -> use the already downloaded assets or use the bundled assets.

IDK yet on how we can enable us to continue downloading the assets if we immediately return the cached–downloaded or bundled–assets. (I'm still not good with threads 😭)

So I have a pretty decent internet speed (`~40mbps`). Maybe it's about the CloudFront server host that sometimes it takes a while to retrieve the assets.

### Demo
![freezing](https://user-images.githubusercontent.com/21978370/93527317-4274f700-f96b-11ea-9faa-92ab93b2480c.gif)

I had a rough and naive benchmark on my end, the visual notification duration would look like:
```
Text 1ms
Visual 2,648ms
Audible 1,154ms

Visual execution duration is 3,842ms
```

<details>
<summary>Logs</summary>
Experienced the freezing twice.

```
2020-09-18 04:41:31,423 [  24135]   WARN - .diagnostic.PerformanceWatcher - UI was frozen for 6490ms...
2020-09-18 04:41:53,217 [  45929]   WARN - .diagnostic.PerformanceWatcher - UI was frozen for 7528ms...
```

```
2020-09-18 04:41:26,382 [  19094]   WARN - tor.plugin.assets.AssetManager - Attempting to download asset https://waifu.assets.unthrottled.io/visuals/greetings/greeting_4.gif 
2020-09-18 04:41:31,423 [  24135]   WARN - .diagnostic.PerformanceWatcher - UI was frozen for 6490ms, details saved to ... 
2020-09-18 04:41:38,323 [  31035]   WARN - tor.plugin.assets.AssetManager - Attempting to download asset https://waifu.assets.unthrottled.io/text/celebration.json 
2020-09-18 04:41:39,369 [  32081]   WARN - tor.plugin.assets.AssetManager - Attempting to download asset https://waifu.assets.unthrottled.io/visuals/dance_crew.gif 
2020-09-18 04:41:41,472 [  34184]   WARN - tor.plugin.assets.AssetManager - Attempting to download asset https://waifu.assets.unthrottled.io/audible/celebration/waoow.mp3 
2020-09-18 04:41:45,691 [  38403]   WARN - tor.plugin.assets.AssetManager - Attempting to download asset https://waifu.assets.unthrottled.io/text/happy.json 
2020-09-18 04:41:46,876 [  39588]   WARN - tor.plugin.assets.AssetManager - Attempting to download asset https://waifu.assets.unthrottled.io/visuals/happy/happy_five.gif 
2020-09-18 04:41:50,769 [  43481]   WARN - tor.plugin.assets.AssetManager - Attempting to download asset https://waifu.assets.unthrottled.io/audible/celebration/tuturuu_mayushi_desu.wav 
2020-09-18 04:41:53,217 [  45929]   WARN - .diagnostic.PerformanceWatcher - UI was frozen for 7528ms, details saved to ... 
2020-09-18 04:42:06,081 [  58793]   WARN - tor.plugin.assets.AssetManager - Attempting to download asset https://waifu.assets.unthrottled.io/visuals/happy/happy_one.gif 
2020-09-18 04:42:18,902 [  71614]   WARN - tor.plugin.assets.AssetManager - Attempting to download asset https://waifu.assets.unthrottled.io/text/smug.json 
2020-09-18 04:42:19,939 [  72651]   WARN - tor.plugin.assets.AssetManager - Attempting to download asset https://waifu.assets.unthrottled.io/visuals/smug/smug_reaction_1.gif 
2020-09-18 04:42:21,386 [  74098]   WARN - tor.plugin.assets.AssetManager - Attempting to download asset https://waifu.assets.unthrottled.io/audible/celebration/sugoi_sugoi.wav 
2020-09-18 04:42:26,034 [  78746]   WARN - tor.plugin.assets.AssetManager - Attempting to download asset https://waifu.assets.unthrottled.io/visuals/smug/smugumin_1.gif 
2020-09-18 04:42:28,590 [  81302]   WARN - tor.plugin.assets.AssetManager - Attempting to download asset https://waifu.assets.unthrottled.io/audible/celebration/padoru_padoru.wav 
2020-09-18 04:42:33,907 [  86619]   WARN - tor.plugin.assets.AssetManager - Attempting to download asset https://waifu.assets.unthrottled.io/visuals/celebration/tohru_approval.gif 
2020-09-18 04:42:35,289 [  88001]   WARN - tor.plugin.assets.AssetManager - Attempting to download asset https://waifu.assets.unthrottled.io/audible/celebration/nice_nice_nice_nice.wav 

```
</details>

**Comment by Unthrottled** [comment]
> We should bundle at least 3 visual notifications that the plugin can fallback into whenever the asset retrieval takes a while

Let's try to avoid bundling things. 

I think this can be mitigated by downloading assets off of the UI thread and on some other thread. We could have a background worker that maintains assets to be downloaded so each event has a local asset to use. We can't wait for the worker thread to download an asset, because it would freeze! 

Either that or we could figure out a non-blocking way to have the UI thread and our worker thread communicate.

Makes things more complicated. However, those are just my initial ideas.

**Comment by Unthrottled** [comment]
> So I have a pretty decent internet speed (~40mbps).

# I'll just put this here 

![image](https://user-images.githubusercontent.com/15972415/93531762-61d93900-f905-11ea-8fc8-8550925f9be6.png)

**Comment by zd-zero** [comment]
Should I say goodbye to my 3rd world country? 😂

**Comment by zd-zero** [comment]
> I think this can be mitigated by downloading assets off of the UI thread and on some other thread. We could have a background worker that maintains assets to be downloaded so each event has a local asset to use.

Right, was thinking if we can have a warming event wherein we download at least 1 to 3 assets for each type. Then, every time an event happens it'll request a process to the background worker so this will prevent downloading all the assets and thus only download another 1 to 3 assets. So, the more the user request events the more assets would be available.

Event 
  -> Request download (background)
  -> Use downloaded asset

**Comment by Unthrottled** [comment]
Cool, I think this is a reasonable thing to fix before releasing `v1.3` WDYT?

Is this something you wanted to work on or did you want to work together on this?

**Comment by zd-zero** [comment]
> Cool, I think this is a reasonable thing to fix before releasing v1.3 WDYT?

Yep, I'll include it in `v1.3` as this will improve the UX.

> Is this something you wanted to work on or did you want to work together on this?

I want to work on this, but I may do it badly lol. So I'll work on it first tomorrow early morning then if I managed to do it and if you have free time you can review it and if there are more things that needed to be changed, let's work on it together.

I plan to finish it tomorrow and hope that we can release it on Sunday.

**Comment by Unthrottled** [comment]
Good Luck, Have Fun!

![GLHF](https://media.giphy.com/media/5BTyQlNh8Ugmc/giphy.gif)

**Comment by Unthrottled** [comment]
@zd-zero 
If you would like some ideas on how to manage things in separate threads, I've solved the same problem on the Doki-Theme. I'm using the same APIs this plugin has, so it should look fairly familiar.

https://github.com/doki-theme/doki-theme-jetbrains/pull/271

**Comment by Unthrottled** [comment]
Closed by #213
